### PR TITLE
WIP Handle Deleted Users

### PR DESCRIPTION
**Pull request by jzila**
r? @patrickxb or @maxtaco

### CODE REVIEW

File: go/libkb/resolve.go [hunk 1]
```diff
@@ -171,6 +171,7 @@ func (r *Resolver) resolveURLViaServerLookup(au AssertionURL, input string, with
 
 	var them *jsonw.Wrapper
 	if them, res.err = ares.Body.AtKey("them").ToArray(); res.err != nil {
+		res.err = NotFoundError{fmt.Sprintf("User not found for query: %q", input)}
```
Comment (maxtaco) [inline go/libkb/resolve.go]: If you're just resetting it, use a temporary `err`?

**Comment by maxtaco** [comment]
small suggestion.  What happens when you `id` from the command line for a deleted user?

File: go/libkb/resolve.go (hunk 1 above)
Comment (jzila) [inline go/libkb/resolve.go]: Yeah I neglected the fact that the typing information is lost over RPC. Hmm...

**Comment by jzila** [comment]
Actually thinking about it, do we even have a way of distinguishing a deleted user from a user that doesn't exist?

**Comment by maxtaco** [comment]
Hmm this change breaks tests...

File: go/libkb/resolve.go (hunk 1 above)
Comment (maxtaco) [inline go/libkb/resolve.go]: I think the right way to check this is right above.

File: go/libkb/loaduser.go [hunk 2]
```diff
@@ -304,7 +304,9 @@ func LoadUserFromServer(g *GlobalContext, uid keybase1.UID, body *jsonw.Wrapper)
 		if err != nil {
 			return nil, err
 		}
-		body = res.Body.AtKey("them")
+		if body, err = res.Body.AtKey("them").ToArray(); err != nil {
```
Comment (patrickxb) [inline go/libkb/loaduser.go]: This seems weird to me.  Doesn't rpc_exim.go already translate SCNotFound to NotFoundError?  

I don't think you'd ever get to this line.

File: go/libkb/resolve.go (hunk 1 above)
Comment (patrickxb) [inline go/libkb/resolve.go]: I don't think I understand the context here.  Is there a case where the API server is returning data for a not found user?  Doesn't line 166 handle this already?

Comment (maxtaco) [inline go/libkb/resolve.go]: @patrickxb  I agree. If it doesn't, it should be changed so it does....

Comment (maxtaco) [inline go/libkb/resolve.go]: And BTW, this is for a **deleted** user which is a slightly different case.

Comment (patrickxb) [inline go/libkb/resolve.go]: Then there will be the DELETED status code returned by user/lookup that we should use instead of looking at json.

Comment (maxtaco) [inline go/libkb/resolve.go]: yeah i agree, does that work?

Comment (maxtaco) [inline go/libkb/resolve.go]: See [this API call](https://keybase.io/_/api/1.0/user/lookup.json?username=thurnish41ey)

Comment (maxtaco) [inline go/libkb/resolve.go]: And also `keybase id thurnish41ey` seems to do the right thing.  Whatever codepath KBFS is using here during rekey might have a bug thogh...

Comment (patrickxb) [inline go/libkb/resolve.go]: We don't have a SCDeleted constant that maps to status_codes.iced DELETED, but it's never come up before.  Is `thurnish41ey` deleted?

Comment (maxtaco) [inline go/libkb/resolve.go]: I thought so...

**Comment by jzila** [comment]
I get:
```
➜ keybase id thurnish41ey
▶ ERROR Not found
➜ keybase id uid:a23de2bbc16771fdfb6ff32fc1d99f19
▶ ERROR User a23de2bbc16771fdfb6ff32fc1d99f19 wasn't found (in Identify2WithUID)
```

**Comment by maxtaco** [comment]
Ok. What do we need to happen here, do we need a new error type for deleted account?

**Comment by jzila** [comment]
Yes, I think that would do it. And then I need to fix KBFS to properly handle that error type

File: go/libkb/loaduser.go (hunk 2 above)
Comment (jzila) [inline go/libkb/loaduser.go]: What do you mean? `AtKey` returns a `go-json.Wrapper`, which has an `err` element of type `go-json.Error`.

Comment (patrickxb) [inline go/libkb/loaduser.go]: I mean there's already an APIError being returned on line 305 if the user isn't found.

Comment (jzila) [inline go/libkb/loaduser.go]: I'm confused as to how in `resolve.go`, the `AtKey("them")` error passes through to KBFS, but here it gets automagically converted to `SCNotFound`.

```
92155 2016-08-03T10:31:17.720242 ▶ [DEBU keybase util.go:387] 1ab0 - resolveURLViaServerLookup(input = "uid:a23de2bbc16771fdfb6ff32fc1d99f19") -> ERROR: <root>.them: no such key: them
92156 2016-08-03T10:31:17.720288 ▶ [DEBU keybase util.go:387] 1ab1 - Resolving full expression "uid:a23de2bbc16771fdfb6ff32fc1d99f19" -> ERROR: <root>.them: no such key: them
92157 2016-08-03T10:31:17.720320 ▶ [DEBU keybase util.go:387] 1ab2 - IdentifyHandler.Resolve2(uid:a23de2bbc16771fdfb6ff32fc1d99f19) -> ERROR: <root>.them: no such key: them
```

**Comment by maxtaco** [comment]
Via that code path, we see [this](https://keybase.io/_/api/1.0/merkle/path.json?poll=10&uid=a23de2bbc16771fdfb6ff32fc1d99f19).

So the API endpoint is giving user deleted...

**Comment by jzila** [comment]
Great. So maybe the right approach is to:
1) Add `SCDeleted_216` to protocol/avdl/constants.avdl
2) Create an `DeletedError{}` in libkb.errors.go
3) Map `SCDeleted` to `DeletedError{}` in libkb/resolve.go
4) Handle `DeletedError{}` in KBFS.

WDYT?

**Comment by maxtaco** [comment]
I agree with 1,2 and 4, but are you sure resolve.go is in play here? It might not be...

**Comment by jzila** [comment]
@maxtaco it's definitely in play, see my log above, namely: `resolveURLViaServerLookup(input = "uid:a23de2bbc16771fdfb6ff32fc1d99f19") -> ERROR: <root>.them: no such key: them`

**Comment by maxtaco** [comment]
For some reason `resolveURLViaServerLookup` doesn't show up in the trace if you run `keybase id` from the command line. Do you know why?

**Comment by jzila** [comment]
It's possible you're using a cached value. `resolveURL` checks the cache first, and only called `resolveURLViaServerLookup` if a cached value is not found.

**Comment by maxtaco** [comment]
makes sense...

**Comment by jzila** [comment]
I'm seeing this:
```
2016-08-23T11:39:53.425112 ▶ [DEBU keybase log.go:30] 0ad - API call merkle/path error: user deleted (error 216)
2016-08-23T11:39:53.425231 ▶ [DEBU keybase log.go:30] 0ae * LoadUser: {Contextified:{g:0xc420154a00} UID:a23de2bbc16771fdfb6ff32fc1d99f19 Name: PublicKeyOptional:false NoCacheResult:false Self:false ForceReload:false AllKeys:false LoginContext:<nil> AbortIfSigchainUnchanged:false ResolveBody:<nil>}: 74.295323ms
...
2016-08-23T11:39:53.426554 ▶ [DEBU keybase log.go:30] 0b6 - Identify2WithUID.Run() -> User a23de2bbc16771fdfb6ff32fc1d99f19 wasn't found (in Identify2WithUID)
2016-08-23T11:39:53.426669 ▶ [DEBU keybase log.go:30] 0b7 - RunEngine(Identify2WithUID) -> ERROR: User a23de2bbc16771fdfb6ff32fc1d99f19 wasn't found (in Identify2WithUID)
2016-08-23T11:39:53.426806 ▶ [DEBU keybase log.go:30] 0b8 - ResolveThenIdentify2::Run -> ERROR: User a23de2bbc16771fdfb6ff32fc1d99f19 wasn't found (in Identify2WithUID)
2016-08-23T11:39:53.426969 ▶ [DEBU keybase log.go:30] 0b9 - RunEngine(ResolveThenIdentify2) -> ERROR: User a23de2bbc16771fdfb6ff32fc1d99f19 wasn't found (in Identify2WithUID)
2016-08-23T11:39:53.427145 ▶ [DEBU keybase log.go:30] 0ba - IdentifyHandler.Identify2 -> ERROR: User a23de2bbc16771fdfb6ff32fc1d99f19 wasn't found (in Identify2WithUID)
2016-08-23T11:39:53.526858 ▶ [DEBU keybase globals.go:313] 0bb Calling shutdown first time through
```

**Comment by jzila** [comment]
OK I told Max in Slack that we can avoid handling deleted users in `Resolve`, and it only needs to be done in Identify, but I now believe it needs to happen in both places. The bug appears in KBFS only via Resolve. In `keybase id` it happens based on the following:
1) `keybase id uid:*` `Resolve` works fine because it short-circuits the actual `resolveURLViaServerLookup` due to `needUsername` not being set.
2) KBFS calls `Resolve2` with `needUsername` set, so the `resolveURLViaServerLookup` issue shows up.
3) `keybase id <username>` repros because it doesn't provide a UID, and thus doesn't short circuit the `resolveURLViaServerLookup` check.

**Comment by patrickxb** [comment]
You need to add it to rpc_exim.go too.

**Comment by jzila** [comment]
Do we have an engine for deleting users? I'm trying to write a test.

**Comment by jzila** [comment]
It appears that my new test is hanging, presumably at `DeleteAccount`. @patrickxb do you have any suggestions?

**Comment by maxtaco** [comment]
does it work locally?

**Comment by jzila** [comment]
I haven't run the tests locally since I don't have a running kbweb on this machine. Once I'm in the office I can try it there. I'm mostly looking for a quick visual through the `DeleteAccount` function to see if my approach to getting the session and calling the API's `delete` are correct.

**Comment by maxtaco** [comment]
I didn't see anything that would cause the test not to complete

File: go/engine/identify2_test.go [hunk 3]
```diff
@@ -665,6 +665,33 @@ func TestIdentify2NoSigchain(t *testing.T) {
...
+// TestIdentifyDeleted tests that identifying a deleted user gives the correct
+// error
+func TestIdentify2Deleted(t *testing.T) {
+	tc := SetupEngineTest(t, "Identify2Deleted")
+	defer tc.Cleanup()
+
+	u := CreateAndSignupFakeUserWithUser(tc, "login")
+	userId := u.User.GetUID()
+	DeleteAccount(tc, u)
+
```
Comment (patrickxb) [inline go/engine/identify2_test.go]: Maybe create a new TestContext here and use it for everything below?  So the id happens in a clean home directory?

There could be some weirdness id'ing yourself after a random API call deletes you...

**Comment by patrickxb** [comment]
I added a comment, but can you send a log of where it is hanging?

File: go/engine/identify2_test.go (hunk 3 above)
Comment (jzila) [inline go/engine/identify2_test.go]: Hmm I did this, no bueno. Figuring out how to get it to output the hang, `kill -ABRT` gives me useless runtime-level output.

Comment (patrickxb) [inline go/engine/identify2_test.go]: KEYBASE_DEBUG=1 go test -run Identify2Deleted

Comment (jzila) [inline go/engine/identify2_test.go]: Great thx, it's getting stuck on this line:
```
2016-08-24T12:50:32.646766 ▶ [DEBU keybase login_state.go:259] 147 - GetPassphraseStreamWithPassphrase() -> ok
2016-08-24T12:50:32.646918 ▶ [DEBU keybase login_state.go:1015] 148 + Account "DeleteAccount"
```

Is the account name significant?

Comment (patrickxb) [inline go/engine/identify2_test.go]: account name shouldn't be significant.

I think we're going to need to do this for real.  Can you create a core ticket?  This should all be in one LoginState request with a session reader on the apiarg, etc.

File: go/engine/identify2_test.go [hunk 4]
```diff
@@ -665,6 +665,36 @@ func TestIdentify2NoSigchain(t *testing.T) {
...
+	i := newIdentify2WithUIDTester(tc.G)
+	tc.G.Services = i
+	arg := &keybase1.Identify2Arg{
+		Uid: userId,
+	}
+	eng := NewIdentify2WithUID(tc.G, arg)
+	ctx := Context{IdentifyUI: i}
+
+	err := eng.Run(&ctx)
```
Comment (patrickxb) [inline go/engine/identify2_test.go]: I'm migrating this test to my branch and I think this is the problem here...you need to call engine.RunEngine(eng, ctx) to run an engine, not the Run method.

**Comment by jzila** [comment]
Patrick has created an engine to achieve this.

**Pull request closed without merging**